crates/file_finder/src/file_finder.rs

```diff
@@ -522,84 +554,100 @@ impl FileFinderDelegate {
     ) {
         if search_id >= self.latest_search_id {
             self.latest_search_id = search_id;
-            let extend_old_matches = self.latest_search_did_cancel
-                && Some(query.path_like.path_query())
-                    == self
-                        .latest_search_query
-                        .as_ref()
-                        .map(|query| query.path_like.path_query());
+            let query_changed = self.query_changed(&query);
+            let extend_old_matches = self.latest_search_did_cancel && !query_changed;
+
+            let selected_match = if query_changed {
+                None
+            } else {
+                self.matches.get(self.selected_index).cloned()
+            };
+
             self.matches.push_new_matches(
                 &self.history_items,
                 self.currently_opened_path.as_ref(),
                 Some(&query),
                 matches.into_iter(),
                 extend_old_matches,
             );
+
+            self.selected_index = selected_match.map_or_else(
+                || self.calculate_selected_index(),
+                |m| self.matches.position(&m).unwrap_or(0),
+            );
+
             self.latest_search_query = Some(query);
             self.latest_search_did_cancel = did_cancel;
-            self.selected_index = self.calculate_selected_index();
+
             cx.notify();
         }
     }
 
+    fn query_changed(&self, new_query: &PathLikeWithPosition<FileSearchQuery>) -> bool {
+        return self.latest_search_query.as_ref().map_or(true, |q| {
```

Comment:
> `return`

Interesting, that Clippy does not complain about this, usually it asks to remove it and `;` and use the expression.
Let's do that at least, but we can go even further and inline the whole method as it's used only once.